src/dbnode/storage/shard.go

```diff
@@ -1409,12 +1409,111 @@ func (s *dbShard) insertSeriesSync(
 	return newEntry, nil
 }
 
+func (s *dbShard) insertNewShardEntriesWithLock(entries []*Entry) {
+	if len(entries) == 0 {
+		return
+	}
+
+	// Fast Path: Check if the entire slice can be appended to the end of the list
+	if s.canAppendToEnd(entries[0]) {
+		s.appendEntriesToEnd(entries)
+		return
+	}
+
+	// If not, proceed with the standard insertion logic
+	elem := s.list.Back()
+	i := len(entries) - 1
+
+	for elem != nil && i >= 0 {
+		currListEntry := elem.Value.(*Entry)
+
+		insertIdx := s.findInsertionIndex(entries, currListEntry.Index, 0, i)
+
+		if insertIdx < len(entries) {
+			i = s.insertEntriesAfter(elem, entries, insertIdx, i)
+		}
+
+		elem = elem.Prev()
+	}
+
+	s.insertRemainingEntriesAtFront(entries, i)
```

Comment:
- once you update the code to work on subslices, you won't need the `i` index being passed as an arg
- remove the `Remaining` bit in the naming.